src/providers/seventv/SeventvBadges.cpp

```diff
@@ -17,61 +18,102 @@ void SeventvBadges::initialize(Settings & /*settings*/, Paths & /*paths*/)
     this->loadSeventvBadges();
 }
 
-boost::optional<EmotePtr> SeventvBadges::getBadge(const UserId &id)
+boost::optional<EmotePtr> SeventvBadges::getBadge(const UserId &id) const
 {
     std::shared_lock lock(this->mutex_);
 
     auto it = this->badgeMap_.find(id.string);
     if (it != this->badgeMap_.end())
     {
-        return this->emotes_[it->second];
+        return it->second;
     }
     return boost::none;
 }
 
-void SeventvBadges::loadSeventvBadges()
+void SeventvBadges::assignBadgeToUser(const QString &badgeID,
+                                      const UserId &userID)
+{
+    const std::unique_lock lock(this->mutex_);
+
+    const auto badgeIt = this->knownBadges_.find(badgeID);
+    if (badgeIt != this->knownBadges_.end())
+    {
+        this->badgeMap_[userID.string] = badgeIt->second;
+    }
+}
+
+void SeventvBadges::clearBadgeFromUser(const QString &badgeID,
+                                       const UserId &userID)
 {
-    // Cosmetics will work differently in v3, until this is ready
-    // we'll use this endpoint.
-    static QUrl url("https://7tv.io/v2/cosmetics");
+    const std::unique_lock lock(this->mutex_);
 
-    static QUrlQuery urlQuery;
-    // valid user_identifier values: "object_id", "twitch_id", "login"
-    urlQuery.addQueryItem("user_identifier", "twitch_id");
+    const auto it = this->badgeMap_.find(userID.string);
+    if (it != this->badgeMap_.end() && it->second->id.string == badgeID)
+    {
+        this->badgeMap_.erase(userID.string);
+    }
+}
 
-    url.setQuery(urlQuery);
+void SeventvBadges::addBadge(const QJsonObject &badgeJson)
+{
+    const auto badgeID = badgeJson["id"].toString();
 
-    NetworkRequest(url)
-        .onSuccess([this](const NetworkResult &result) -> Outcome {
-            auto root = result.parseJson();
+    const std::unique_lock lock(this->mutex_);
 
+    if (this->knownBadges_.find(badgeID) != this->knownBadges_.end())
+    {
+        return;
+    }
+
+    auto emote = Emote{
+        .name = EmoteName{},
+        .images = SeventvEmotes::createImageSet(badgeJson),
+        .tooltip = Tooltip{badgeJson["tooltip"].toString()},
+        .homePage = Url{},
+        .id = EmoteId{badgeID},
+    };
+
+    if (emote.images.getImage1()->isEmpty())
+    {
+        return;  // Bad images
+    }
+
+    this->knownBadges_[badgeID] =
+        std::make_shared<const Emote>(std::move(emote));
+}
+
+void SeventvBadges::loadSeventvBadges()
```

Comment:
Given that the endpoint is being removed in November, I would rather just rip it out here and be done with it